matRad_SFUDoptimization.m

```diff
@@ -0,0 +1,136 @@
+%   cst:         matRad cst struct
+%   dij:         matRad dij struct (optional)
+%                  
+%   ct:          matRad ct struct (optional, only needed if no dij provided)
+%   stf:         matRad stf struct (optional, only if needed no dij provided)
+%
+% output
+%   resultGUI:  struct containing optimized fluence vector, dose, and (for
+%               biological optimization) RBE-weighted dose etc.
+%   (info:       struct containing information about optimization)
+%
+% References
+%   [1]    https://ro-journal.biomedcentral.com/articles/10.1186/s13014-016-0705-8
+%
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+% adjust cst for single beams
+sb_cst = cst;
+for i=1:size(sb_cst,1)
+    for j = 1:size(sb_cst{i,6},1)
+        % biological dose splitting
+        if isfield(dij,'mAlphaDose') && isfield(dij,'mSqrtBetaDose')
+            ab = sb_cst{i,5}.alphaX / sb_cst{i,5}.betaX;
+            % dose per fraction
+            fx_dose = sb_cst{i,6}(j).dose / pln.numOfFractions;
+            % calculate dose per beam per fraction according to [1]
+            fx_dose = -0.5*ab +sqrt( 0.25*ab^2 + ...
+                fx_dose/pln.numOfBeams *(fx_dose + ab));
+            % calculate pseudo total Dose per Beam
+            sb_cst{i,6}(j).dose = fx_dose * pln.numOfFractions;
+            
+        % physical dose splitting
+        else
+            sb_cst{i,6}(j).dose = sb_cst{i,6}(j).dose/pln.numOfBeams;
+        end
+    end
+end
+
+if ~isempty(dij)
+    % calculate dij with total dij being present
+        
+    % initialise total weight vector
+    wTot = zeros(dij.totalNumOfBixels,1);
+
+    for i = 1:pln.numOfBeams
+        % columns in total dij for single beam
+        sb_col = find(dij.beamNum == i);
+        % construct dij for single beam
+        sb_dij.numOfBeams = 1;
+        sb_dij.numOfVoxels = dij.numOfVoxels;
+        sb_dij.resolution = dij.resolution;
+        sb_dij.numOfRaysPerBeam = dij.numOfRaysPerBeam(i);
+        sb_dij.totalNumOfRays = sb_dij.numOfRaysPerBeam;
+        sb_dij.totalNumOfBixels = size(sb_col, 1);
+        sb_dij.dimensions = dij.dimensions;
+        sb_dij.numOfScenarios = dij.numOfScenarios;
+        sb_dij.bixelNum = dij.bixelNum(sb_col);
+        sb_dij.rayNum = dij.rayNum(sb_col);
+        sb_dij.beamNum = dij.beamNum(sb_col);
+        sb_dij.physicalDose{1} = dij.physicalDose{1}(:, sb_col);
+        if isfield(dij, 'RBE')
+            sb_dij.RBE = dij.RBE;
+        end
+        if isfield(dij, 'mLETDose')
+            sb_dij.mLETDose = dij.mLETDose(:, sb_col);
+        end
+        if isfield(dij,'mAlphaDose') && isfield(dij,'mSqrtBetaDose')
+            sb_dij.mAlphaDose{1} = dij.mAlphaDose{1}(:, sb_col);
+            sb_dij.mSqrtBetaDose{1} = dij.mSqrtBetaDose{1}(:, sb_col);
+        end
+
+        % adjust pln to one beam only
+        sb_pln = pln;
+        sb_pln.gantryAngles = pln.gantryAngles(i);
+        sb_pln.couchAngles = pln.couchAngles(i);
+
```

Comment:
What about sb_pln.numOfBeams? Maybe not needed but should better be included if we make changes to the optimization to use this parameter at a later point...